server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java

```diff
@@ -673,128 +706,127 @@ public CompletableFuture<Revision> updateWriteQuota(
-            }
-            return findPerRolePermissions(repositoryMetadata, registration.role());
+            final TokenRegistration projectTokenRegistration = metadata.tokens().get(appId);
+            final ProjectRole projectRole = projectTokenRegistration != null ? projectTokenRegistration.role()
+                                                                             : ProjectRole.GUEST;
+            return repositoryRole(roles, tokenRepositoryRole, projectRole);
         });
     }
 
-    /**
-     * Finds {@link Permission}s which belong to the specified {@link User} from the specified
-     * {@code repoName} in the specified {@code projectName}.
-     */
-    private CompletableFuture<Collection<Permission>> findPermissions0(String projectName, String repoName,
-                                                                       User user) {
+    private CompletableFuture<RepositoryRole> findRepositoryRole0(String projectName, String repoName,
+                                                                  User user) {
         requireNonNull(projectName, "projectName");
         requireNonNull(repoName, "repoName");
         requireNonNull(user, "user");
 
         return getProject(projectName).thenApply(metadata -> {
             final RepositoryMetadata repositoryMetadata = metadata.repo(repoName);
-            final Member member = metadata.memberOrDefault(user.id(), null);
+            final Roles roles = repositoryMetadata.roles();
+            final RepositoryRole userRepositoryRole = roles.users().get(user.id());
 
-            // If the member is guest.
-            if (member == null) {
-                return repositoryMetadata.perRolePermissions().guest();
-            }
-            final Collection<Permission> p = repositoryMetadata.perUserPermissions().get(member.id());
-            if (p != null) {
-                return p;
-            }
-            return findPerRolePermissions(repositoryMetadata, member.role());
+            final Member projectUser = metadata.memberOrDefault(user.id(), null);
+            final ProjectRole projectRole = projectUser != null ? projectUser.role() : ProjectRole.GUEST;
+            return repositoryRole(roles, userRepositoryRole, projectRole);
         });
     }
 
-    private static Collection<Permission> findPerRolePermissions(RepositoryMetadata repositoryMetadata,
-                                                                 ProjectRole role) {
-        switch (role) {
-            case OWNER:
-                return repositoryMetadata.perRolePermissions().owner();
-            case MEMBER:
-                return repositoryMetadata.perRolePermissions().member();
-            default:
-                return repositoryMetadata.perRolePermissions().guest();
+    @Nullable
+    private static RepositoryRole repositoryRole(Roles roles, @Nullable RepositoryRole repositoryRole,
+                                                 @Nullable ProjectRole projectRole) {
```

Comment:
`projectRole` seems not null always in the current usage.
```suggestion
                                                 ProjectRole projectRole) {
```